src/core/highlight.ts

```diff
@@ -24,32 +40,31 @@ import type { Grammar, GrammarToken, GrammarTokens, RegExpLike } from '../types'
 export function highlight (
 	this: Prism,
 	text: string,
-	language: string,
+	languageRef: string | LanguageLike,
 	options: HighlightOptions = {}
 ): string {
 	const prism = this ?? singleton;
 
-	const languageId = prism.components.resolveAlias(language);
-	const grammar = options.grammar ?? prism.languageRegistry.getLanguage(languageId);
+	const { id, def, language } = prism.languageRegistry.resolveRef(languageRef);
 
-	const env: Record<string, any> = {
+	let env: HookEnv['after-tokenize'] = {
 		code: text,
-		grammar,
+		languageId: id,
+		languageDef: def,
 		language,
 	};
 
 	prism.hooks.run('before-tokenize', env);
 
-	if (!env.grammar) {
-		throw new Error('The language "' + env.language + '" has no grammar.');
+	if (!env.language) {
+		throw new Error(`No language definition found for ${env.languageId}.`);
 	}
 
-	env.tokens = prism.tokenize(env.code, env.grammar.resolvedGrammar);
+	env = env as unknown as HookEnv['after-tokenize'];
+	env.tokens = tokenize.call(prism, env.code, env.language!.resolvedGrammar);
```

Comment:
Do we really need to use `call` here? `tokenize` already expects a Prism instance as the first argument. Or do you have some changes to its signature in mind?

Reply:
Or probably it's a TIL thing: if we call the first argument of a function `this`, we can use it inside the function body to assign some default context the function will work with? And to pass this context explicitly, we use `call`?

Reply:
It's definitely confusing syntax though!